fw/http.c

```diff
@@ -5758,6 +5730,66 @@ tfw_h1_req_process(TfwStream *stream, struct sk_buff *skb)
 	return hmsib;
 }
 
+static void
+extract_req_host(TfwHttpReq *req)
+{
+	int hid = 0;
+	TfwStr *hdrs = req->h_tbl->tbl;
+
+	if (TFW_MSG_H2(req)) {
+		/* RFC 9113, sec-8.3.1:
+		 * The recipient of an HTTP/2 request MUST NOT use
+		 * the Host header field to determine the target
+		 * URI if ":authority" is present.*/
+		if (!TFW_STR_EMPTY(&hdrs[TFW_HTTP_HDR_H2_AUTHORITY]))
+			hid = TFW_HTTP_HDR_H2_AUTHORITY;
+		else
+			hid = TFW_HTTP_HDR_HOST;
+		__h2_msg_hdr_val(&hdrs[hid], &req->host);
+	} else {
+		/* req->host can be only filled by HTTP/1.x parser from
+		 * absoluteURI, so we act as described by RFC 9112, sec-3.2.2
+		 * (https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2.2):
+		 * When an origin server receives a request with an
+		 * absolute-form of request-target, the origin server
+		 * MUST ignore the received Host header field (if any)
+		 * and instead use the host information of the request-target.
+		 */
+		if (TFW_STR_EMPTY(&req->host)) {
+			hid = TFW_HTTP_HDR_HOST;
+			tfw_http_msg_clnthdr_val(req, &hdrs[TFW_HTTP_HDR_HOST],
+						 TFW_HTTP_HDR_HOST,
+						 &req->host);
+		}
+	}
+}
+
+static bool
+validate_authority_rfc9112(TfwHttpReq *req)
+{
+	/* https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2.2
+	 * A client MUST send a Host header field in an HTTP/1.1
+	 * request even if the request-target is in the absolute-form
+	 */
+	return req->uri_host ||
+		!TFW_STR_EMPTY(&req->h_tbl->tbl[TFW_HTTP_HDR_HOST]);
+}
+
+static bool
+validate_authority_rfc9113(TfwHttpReq *req)
+{
+	/* https://datatracker.ietf.org/doc/html/rfc9113#section-8.3.1
+	 * A server SHOULD treat a request as malformed if it contains
+	 * a Host header field that identifies an entity that differs
+	 * from the entity in the ":authority" pseudo-header field.
+	 * */
+	TfwStr authority, host;
+	__h2_msg_hdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_H2_AUTHORITY], &authority);
+	__h2_msg_hdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_HOST], &host);
+	return TFW_STR_EMPTY(&authority) || TFW_STR_EMPTY(&host)
+		|| tfw_strcmp(&authority, &host) == 0;
+}
+
```

Comment:
Let's do these validations in `extract_req_host()`:
1. `tfw_http_req_process()` is already too big
2. there is no sense to check the request version twice and spread the logic among many places (we already have the frang check)